### ISSUE TITLE
1.5.1 does not build on non-windows

### ISSUE THREAD

**Issue by Zirias** [jpage8580/GTUltra#57]
I'm just trying to upgrade the FreeBSD port of GTUltra, unfortunately I can't get it to build. The errors seem all related to `gt2reloc.c`.

The first issue is usage of `keybd_event()`, which is a Windows-specific function. I don't really understand the purpose of "faking" keystrokes here, but just removing it for non-Windows helps:

```
@@ -219,10 +219,12 @@ int main(int argc, char **argv)
        }
        else {
                usage();
+#ifdef __WIN32__
                // ENTER key down
                keybd_event(VK_RETURN, 0x9C, 0, 0);
                // ENTER key up
                keybd_event(VK_RETURN, 0x9C, KEYEVENTF_KEYUP, 0);
+#endif
                exit(-1);
 
        }
@@ -452,10 +454,12 @@ int main(int argc, char **argv)
        // perform relocation
        relocator(&gtObject, 1);
 
+#ifdef __WIN32__
        // ENTER key down
        keybd_event(VK_RETURN, 0x9C, 0, 0);
        // ENTER key up
        keybd_event(VK_RETURN, 0x9C, KEYEVENTF_KEYUP, 0);
+#endif
        // Exit
        exit(0);
        return 0;
```

The next issue is a linker complaint about double definition of `gtObject`. I have to really start guessing here, my guess is it *should* reference an object defined somwhere else, so the following patch would help if this assumption would be correct:

```
@@ -95,7 +95,7 @@ int useOriginalGTFunctionKeys = 0;
 char debugTextbuffer[MAX_PATHNAME];
 
 int useOriginalGTFunctionKeys = 0;
-GTOBJECT gtObject;
+extern GTOBJECT gtObject;
 int useRepeatsWhenCompressing = 1;
 char infoTextBuffer[256];
 int midiEnabled = 0;
```

But then, I really have no idea about the next error, linking `gt2reloc` complains about an undefined symbol `playUntilEnd`, referenced by `gorder.o`. This one *is* defined in `gt2stereo.o`, but linking that as well creates a whole list of duplicate symbols instead.

I'll hold back the update of the FreeBSD port for now and hope for some feedback, because I don't want to build/package a broken version of GTUltra ... TIA!

**Comment by jpage8580** [comment]
Thanks for this. And - sorry!

Have pushed changes. Can you try again now?

**Comment by Zirias** [comment]
Thanks a lot (and that was really quick!), pulling in your latest commit as a patch makes the port build again!

So, I'll continue working on updating the port, 1.5.1 will probably be committed to FreeBSD's ports collection later today.

**Comment by jpage8580** [comment]
Great!
The keyboard code - All I wanted was to do a printf!! But SDL2 doesn't
allow that so I had to use their debug log commands to get anything to
display. However, that then sets things up in a way that requires return to
be pressed to quit.. Use case was someone who wanted this called in a
batch, repeatedly - hence my hacks!

**Comment by jpage8580** [comment]
I don't see any of those warnings. My build options hide them from me - But
jeeeezzz, they're shocking! haha!
The issue is due to changes in SDL2 - It worked fine in SDL!

**Comment by Zirias** [comment]
I like to enable all possible warnings when coding in C or C++ (like, `-Wall -Wextra -pedantic`) and even test with both `clang` and `gcc`. Although there are many "false positives", getting your code free of warnings really helps avoid actual bugs ...

But anyways, all build tests are fine now, so FreeBSD has the latest GTUltra version now :+1: 
https://www.freshports.org/audio/gtultra/
https://cgit.freebsd.org/ports/commit/?id=083957ec3fa82fef7b934b50e6dbec384119b253

If only every upstream would fix issues THAT quickly, haha!

**Comment by jpage8580** [comment]
Thanks so much for doing this

**Comment by Zirias** [comment]
TBH, I have to thank you.

I see this code still has some quality issues (this wall of warnings is unsettling, I won't promise anything, but if I find some time, I might have a look), but I guess most of it is inherited from original goattracker ...

Nevertheless: My personal background is, I'm one of those "C64 kids", I'm sometimes a bit active in the scene as a coder, I even composed music (using trackers as well as my own player code ...) – and at the same time, I'm a "professional" software dev and architect and just love FreeBSD (and made it to a ports committer there quickly).

So, my very personal goal is to make FreeBSD a top notch platform for C64 enthusiasts. And of course, that's only possible with free and open-source software available (that e.g. also allows binary redistribution, users don't want to compile themselves!). Unfortunately, some very useful tools aren't free (SpritePad, CharPad), some others are "free" but not open-source (DirMaster), so you can't build and package them for FreeBSD. Concerning D64 editing, I tried to fill the gap myself with "v1541commander", and I should really get back working on it! But hey, always too many private projects ;)

Long story short, I'm very thankful for any free and open-source software around the C64 I can find!

**Comment by Zirias** [comment]
Regression in the commit to fix warnings, see https://github.com/jpage8580/GTUltra/commit/19132bcba908fed2820f54d90e3ec2bfc679f506#diff-56a9aba3cf01edbcf0b5e96f74054c6c9e60b65c72db1cc04064573e854e40ddL458

Result: It again doesn't build on non-windows systems :(

I would really like to understand what exact problem triggered that "workaround" ... IMHO it should be analyzed and properly solved, MAYBE I can help if I find time this weekend ....

**Comment by jpage8580** [comment]
I've no idea how that regressed!
Have fixed...again...hopefully

This basically explains why I bodged the code...

The original GTreloc (using SDL) has this code:

#ifdef __WIN32__
  /*
    SDL_Init() reroutes stdout and stderr, either to stdout.txt and
stderr.txt
    or to nirwana. simply reopening these handles does, other than
suggested on
    some web pages, not work reliably - opening new files on CON using
different
    handles however does.
  */
  STDOUT = fopen("CON", "w");
  STDERR = fopen("CON", "w");
#endif


This no longer works with SDL2, so I couldn't find any way to get output to
the cmd window.

see here:
https://stackoverflow.com/questions/14593783/sdl-console-output-works-when-debuging-but-not-when-run-with-the-exe

The only way I could get anything to output was to use SDL_Log.
However, using this will leave the cmd window in a state where it wants the
used to press return to exit that mode..
This wasn't good for a user case where someone was using gt2reloc in a
batch file. So, I basically just added a hack to force the return code!

**Comment by Zirias** [comment]
Ahh thanks for both the quick fix and some context about the underlying problem! :+1:

As for updating the FreeBSD port, I guess I will just delay it a bit for now (but of course, still check the current tip of the main branch!)

Regarding the issue with SDL.... I hope I will find some time to have a look myself. There just must be some sane way to solve it :open_mouth:

**Comment by jpage8580** [comment]
Indeed! My hack was literally to get a version that could be used asap. I
could only find ways to solve this that involved recompiling SDL2, which I
didn’t want to do.

Silly thing is, the problem only occurs due to gt2reloc using the same code
base as GT - even though the actual code to do the export doesn’t really
need it. I could just copy the relevant structures and variables into
gt2reloc and build it as standalone. But that might then involve a lot more
upkeep, keeping both in sync.

**Comment by Zirias** [comment]
> Silly thing is, the problem only occurs due to gt2reloc using the same code base as GT - even though the actual code to do the export doesn’t really need it. I could just copy the relevant structures and variables into gt2reloc and build it as standalone. But that might then involve a lot more upkeep, keeping both in sync.

Yeah about this ... and maybe that's indeed the way to happiness ... I have already noticed that the whole codebase could need some sane "structuring".

You certainly would not want to duplicate code, but you **would** want to separate code needing SDL from code that doesn't (and just use the linker ....), at the very least.

But that's probably a **huge** amount of work, dissecting the existing code and refactor it into modules (compilation units) that just have one job, as little dependencies as possible, and make sense...

**Comment by Zirias** [comment]
After the latest commit, it [builds fine again](https://home.palmen-it.de/testbuilder/build.html?mastername=14a-default&build=2023-05-26_21h03m09s), so closing this again, and thanks for that!

**Comment by jpage8580** [comment]
Yeah, I’ve already spent substantial time on defining structures to allow
for multiple instances of the player code, etc. Wasnt fun, but totally
necessary.
Maybe worth moving all of those structs into a single header so that both
GT editor and reloc could just use that.

**Comment by jpage8580** [comment]
Had a good look at removing sdl from gt2reloc last night. It’s everywhere
though, as it’s required by the bme library. Not worth the effort just for
that stand-alone relocator that isn’t really that necessary anyway
